### PR TITLE
FS-4008: Full application comments

### PR DESCRIPTION
**Pull request by hamzabinkhalid**
### Ticket
https://dluhcdigital.atlassian.net/browse/FS-4008

### Change description
- govuk-frontend version downgraded from 5.3.0 to 4.7.0 beacause 5.3 breaks some frontend components
- Added and used new comment type enum WHOLE_APPLICATION
- Comment section added on the assessor_tasklist page
- New Comment section made scrollable

#### Comment Section:

Screenshot 2024-04-24 at 08 48 22

#### Add new Comment:

Screenshot 2024-04-24 at 08 49 28

Screenshot 2024-04-24 at 08 50 23

#### Edit a comment:

Screenshot 2024-04-24 at 08 51 07

Screenshot 2024-04-24 at 08 51 29

#### Comment history:

Screenshot 2024-04-24 at 08 52 05

### CODE REVIEW

File: app/blueprints/assessments/routes.py [hunk 1]
```diff
@@ -680,6 +683,83 @@ def application(application_id):
...
+    add_comment_argument = request.args.get("add_comment") == "1"
+    edit_comment_argument = request.args.get("edit_comment") == "1"
+    comment_form = CommentsForm()
+
+    if add_comment_argument and comment_form.validate_on_submit():
+        comment = comment_form.comment.data
+        submit_comment(
+            comment=comment,
+            application_id=application_id,
+            sub_criteria_id="",
```
Comment (tferns) [inline app/blueprints/assessments/routes.py:695]: Might be worth adding a comment in the code here noting why sub criteria is blank?
```suggestion
            sub_criteria_id="",  # No subcriteria id indicates it belongs to the entire application.
```

Or something like that?

File: app/blueprints/assessments/routes.py [hunk 2]
```diff
@@ -680,6 +683,83 @@ def application(application_id):
...
+            user_id=g.account_id,
+            theme_id="",
```
Comment (tferns) [inline app/blueprints/assessments/routes.py:697]: Maybe same with `theme_id`?

File: app/blueprints/assessments/routes.py [hunk 3]
```diff
@@ -680,6 +683,83 @@ def application(application_id):
...
+        theme_id="",
+        comment_type=CommentType.WHOLE_APPLICATION.name,
+    )
+
+    # TODO add test for this function in data_operations
+    theme_matched_comments = (
+        match_comment_to_theme(comment_response, "", state.fund_short_name)
+        if comment_response
+        else None
+    )
```
Comment (tferns) [inline app/blueprints/assessments/routes.py:724]: Does the `match_comment_to_theme` function just work with an empty second argument I'm guessing?

Might be worth using keyword arguments if we're using a blank string, so it's more clean, something like this maybe:

```suggestion
    # TODO add test for this function in data_operations
    theme_matched_comments = None
    if comment_response:
        theme_matched_comments = match_comment_to_theme(
            comment_response=comment_response,
            themes="",
            fund_short_name=state.fund_short_name
        )
```

After looking at that, it looks like second argument is a list and not a string, should maybe we provide an empty list?  

https://github.com/communitiesuk/funding-service-design-assessment/blob/39cf96a/app/blueprints/services/data_services.py#L669

File: app/blueprints/assessments/templates/assessor_tasklist.html [hunk 4]
```diff
@@ -147,41 +150,69 @@
...
-            {{ assessment_complete(state, form.csrf_token, application_id)}}
+    {% if sub_criteria_status_completed and g.access_controller.has_any_assessor_role and (not is_qa_complete) and (not flag_status) and not is_expression_of_interest %}
+        {{ assessment_complete(state, form.csrf_token, application_id)}}
+    {% endif %}
+</div>
+</div>
+<div class="govuk-grid-row">
+    <div class="govuk-grid-column-one-third">
+        {{ comment(comments) }}
+        {% if display_comment_box != True and not display_comment_edit_box %}
```
Comment (tferns) [inline app/blueprints/assessments/templates/assessor_tasklist.html:161]:
```suggestion
        {% if not display_comment_box and not display_comment_edit_box %}
```

More pythonic and consistent, but a nitpick so feel free to ignore

File: app/blueprints/assessments/templates/assessor_tasklist.html [hunk 5]
```diff
@@ -147,41 +150,69 @@
...
+                    add_comment="1",
+                    _anchor="comment"
+                ) }}">
+                {% if comments == None %} Add a comment
+                {% else %} Add another comment
+                {% endif %}
+        {# djlint:off #}</a>{# djlint:on #}
+        {% endif %}
+
+        {% if display_comment_box == True %}
```
Comment (tferns) [inline app/blueprints/assessments/templates/assessor_tasklist.html:179]:
```suggestion
        {% if display_comment_box %}
```

nitpick, feel free to ignore

File: app/blueprints/assessments/templates/macros/comments.html [hunk 6]
```diff
@@ -1,69 +1,89 @@
...
-                            sub_criteria_id=comment.sub_criteria_id,
-                            theme_id=comment.theme_id,
-                            show_comment_history="1",
-                            comment_id=comment.id) }}">See history</a>
+            {% if comments == None %}
+            <div class="govuk-body govuk-!-margin-top-6">
+                No comments have been left yet.
+                </div>
+            {% else %}
+            {% if '' in comments %}
```
Comment (tferns) [inline app/blueprints/assessments/templates/macros/comments.html:12]: What is this doing, checking if a blank comment exists in comments?  Big strange from a glance

**Review by tferns — COMMENTED** [review]
A few comments!  A little confused with the blank string representing no theme or no comments, but I think I get the gist of it!  Nice work 👍 

Let me know if you want me to approve after reviewing my comments

**Review by tferns — DISMISSED** [review]

File: app/blueprints/assessments/templates/assessor_tasklist.html (hunk 5 above)
Comment (hamzabinkhalid) [reply app/blueprints/assessments/templates/assessor_tasklist.html]: fixed

File: app/blueprints/assessments/routes.py (hunk 3 above)
Comment (hamzabinkhalid) [reply app/blueprints/assessments/routes.py]: fixed, added keyword arguments

File: app/blueprints/assessments/templates/assessor_tasklist.html (hunk 4 above)
Comment (hamzabinkhalid) [reply app/blueprints/assessments/templates/assessor_tasklist.html]: good shout, resolved.

**Review by srh-sloan — APPROVED** [review]
Looks good, well done 👍

**Pull request merged**